Adding CredentialType and accompanying documentation for re-using cards.

peteward opened the pull request:
Firstly, a big thank you for writing this library, it saved me so much time ❤️ 

This adds a missing element around re-using cards, especially now that 3D Secure V2 is mandatory.

I wasn't quite sure about my added documentation as it potentially conflicts with a section around `Repeat Payments`. Instead of replacing this I added my own section.

peteward commented on src/Request/Model/ReusableCard.php:
```diff
@@ -24,10 +24,6 @@ class ReusableCard extends AbstractCard
     {
         $this->cardIdentifier = (string)$cardIdentifier;
-
-        if (isset($save)) {
```
unrelated, but this looked like it shouldn't be here...

mar7t1n commented:
Hoping you could help. Since the end of Feb 2022, the majority of our repeat transactions now get declined with "The Authorisation was Declined by the bank. SCA required.". The original transactions all had v2 applied and a credentialType sent. Opayo don't seem to know what's wrong and just pass us back to the API. Are we just sending the wrong data, or have we lost the ability to action a repeat and not been told how to renew it.

judgej commented:
Thanks, no sure why I missed this PR.

Just to note, the method that constructs the URL will automatically prefix the resource with a slash if it does not already have one:

https://github.com/academe/opayo-pi/blob/master/src/Model/Endpoint.php#L106

I can't remember why it is done like this (eight years ago), but I'm guessing there is a reason. If the "/" is removed between {version} and {resource} in the URL template, then it will work with or without a slash beinf provided, and won't break sites that may already provide it.

peteward commented:
Yeah good point.

I just updated this quickly to fix a broken site yesterday, I'll will remove the slashes to avoid the conflict.

The main issue though was not the legacy domain URL but that Opayo / Elavon now returns datetime strings with more decimal places than `DateTime()` can handle and it blew up. I added a very hacky truncating of this time to 3 decimal places but it did the job.

peteward commented:
Updated for latest master branch.

This code has been in use for 2 years for managing saved cards so I'm confident that it works 😄

peteward commented:
It's an embarrassingly old 7.4 sadly...

mar7t1n commented:
Users should be aware that Opayo now Elavon charge for holding card tokens for repeat use. My client was being billed £300 a month for having tokens. Which in fact we aren't using because so many fell out that it became easier to just tell customers they needed to pay their balance interactively when they were ready to do so. If your Tx volume is small you are probably better off with Stripe.com, if it's large, then repeat payments is going to cost you a small fortune unless you only need to hold them for a short period.

judgej commented:
Just checked the date parsing of '2024-03-01T04:29:31.501675271Z' in PHP 7.3 and 8.0.

* 7.3 returns null (it may raise an error too, but I'm not seeing that in `php -a`)
* 8.0 returns the correct DateTime.

I think a longer time solution would be to attempt to parse with no manipulation first, then if no success, try removing the microseconds (leaving the milliseconds). Some frameworks may divert parsing errors into exceptions, so it would need to catch those too. In the meantime, we can go with what we have just to get it working for older PHP versions.

I also checked my account for date formats being returned, and did not see the microseconds. It may just be some responses I have not looked at, or may have been transient. It's all a bit random and unpredictable.

judgej commented:
I wonder, could we change the truncation to microseconds, since they do work with PHP 7.3 at least?

So if the part after the "." matches `/^[0-9]{6}[0-9]+Z$/` (six digits plus at least one more) then take the first 6 digits and append the 'Z'?

judgej commented:
Added a few bits on PR #5 - support micro-seconds but truncate any fractions of that. Does this work for you?

peteward commented:
Sounds great 👍

Sorry, the date fix wasn't very well thought through, it was just enough to get a site back up

judgej commented:
Oh, it was perfect, thank you. I'm just a digger of edge cases and wanted to see where the limits were. All your changes are included.
